common/src/main/java/io/pravega/common/io/DirectDataOutput.java

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.io;
+
+import io.pravega.common.util.BufferView;
+import java.io.IOException;
+
+/**
+ * Defines an object meant for serializing data (usually an {@link java.io.OutputStream}) that can write various inputs
+ * more efficiently than otherwise (i.e., byte-by-byte copy).
+ */
+public interface DirectDataOutput {
```

Comment:
Several reasons:
1. It has `writeBuffer(BufferView)`.
2. I do not have to reimplement all `DataOutput` methods; I only need to do the ones for which I have proper optimizations in place. Plus, every other data type can be reduced to one of these here.
3. `DataOutputStream`, which is the sole class implementing `DataOutput`, made all of its implementation methods `final` so I cannot subclass that method to provide more efficient implementations for them. 

I could see that in a future iteration, we can clean this up and provide a uniform implementation, but for now I did not want to force every class implementing this (`ByteBufferOutputStream`, `DataFrameOutputStream`) to subclass `DataOutputStream` (for reason #3 above) or to have to independently reimplement the methods in `DataOutputStream`.

Reply:
DataOutput and DataOutputStream are different.

Reply:
Yes. I did not want to implement all methods defined on DataOutput. I only care about a select few. I brought up DataOutputStream because it is the encoding format that we want to follow. My only choice would be to reimplement that everywhere I implement this interface, which is not my intention.